src/pi-src/data_collection.cpp

```diff
@@ -104,14 +142,153 @@ void disp_ping(ping p){
+                snd = &(ping_collectors[pin].pings[j]);
+
+                if(pings_match(*fst, *snd)){
+                    /* Pings found! Return in chronological order */
+                    int in_order = fst->start < snd->start;
+                    *first  = in_order ? fst : snd;
+                    *second = in_order ? snd : fst;
+                    return id;
+                }
+            }
+        }
+    }
+    return -1; /* failure */
+}
+
+/**
+ * Finds two matching pins on a given pin where both pings encode some ID.
+ * @param pin    - The pin [0..3] to read from
+ * @param first  - The chronologically-first ping will be pointed at by first
+ * @param second - The chronologically-second ping will be pointed at by second
+ * @returns      - The ID of the pings, if found, or -1 if no match found
+ */
+int find_match_on_pin(int pin, ping **first, ping **second){
+    ping *fst;
+    ping *snd;
+    int candidate_id = -1;
+    for(int i=0; i<PING_BUF_LEN; i++){
+        // TODO: skip if id < 0?
+        /* Check each ping, one at a time */
+        fst = &(ping_collectors[pin].pings[i]);
+        candidate_id = id_decode_ping(*fst);
+
+        /* Compare this ping with one that comes afterwards */
+        for(int j=i+1; j<PING_BUF_LEN; j++){
+            snd = &(ping_collectors[pin].pings[j]);
+
+            /* If pings have same ID and delay is correct, we found a match! */
+            if(pings_match(*fst, *snd)){
+                int in_order = fst->start < snd->start;
+                *first  = in_order ? fst : snd;
+                *second = in_order ? snd : fst;
+                return candidate_id;
+            }
+        }
+    }
+    return -1;
+}
+
+/**
+ * When a ping has been used, zero it out so that it is not used twice.
+ */
+int clear_ping(ping *p){
+    int pin = p->pin;
+    int i = 0;
+    while(i<PING_BUF_LEN && &(ping_collectors[pin].pings[i]) != p){ i++; }
+    if(i<PING_BUF_LEN){
+        memset(p, 0, sizeof(ping));
+        return 1;
+    }
+    return 0;
+}
+
+/**
+ * Look for a total of 8 pings that can be matched to a single iCRAB broadcast
+ * @param set - Pointers to the appropriate pings will be stored here
+ * @returns   - The ID of the crab, if enough pings found, else -1
+ */
+int get_set(full_set *set){
+    ping *f, *s;
+    int id;
+    /* Find pings on pin A */
+    if((id = find_match_on_pin(0, &f, &s)) < 0) return -1;
+    set->pings_a[0] = f;
+    set->pings_a[1] = s;
+    /* Find pings on pin B */
+    if(find_match_on_pin_with_id(1, id, &f, &s) != id) return -1;
+    set->pings_b[0] = f;
+    set->pings_b[1] = s;
+    /* Find pings on pin C */
+    if(find_match_on_pin_with_id(2, id, &f, &s) != id) return -1;
+    set->pings_c[0] = f;
+    set->pings_c[1] = s;
+    /* Find pings on pin D */
+    if(find_match_on_pin_with_id(3, id, &f, &s) != id) return -1;
+    set->pings_d[0] = f;
+    set->pings_d[1] = s;
+    return id;
+}
+
+/**
+ * Deletes ping entries for their buffers.
```

Comment:
"for their buffers" or "from their buffers"?

Reply:
Indeed! Thanks